src/ostree/ot-builtin-commit.c

```diff
@@ -328,6 +334,55 @@ parse_keyvalue_strings (char             **strings,
   return ret;
 }
 
+#ifdef OSTREE_ENABLE_EXPERIMENTAL_API
+static gboolean
+fill_collection_id (OstreeRepo    *repo,
+                    GVariant      *metadata,
+                    GVariant     **out_metadata,
+                    GCancellable  *cancellable,
+                    GError       **error)
+{
+  if (opt_orphan)
+    {
+      if (opt_collection_id)
+        {
+          g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+                       "Can't use --collection-id together with --orphan");
+          return FALSE;
+        }
+      *out_metadata = (metadata != NULL) ? g_variant_ref (metadata) : NULL;
+      return TRUE;
+    }
+
+  g_assert_nonnull (opt_branch);
+
+  const char *collection_id = opt_collection_id;
+  const char *repo_collection_id = ostree_repo_get_collection_id (repo);
+  if (collection_id == NULL)
+    collection_id = repo_collection_id;
+  else if (g_strcmp0 (collection_id, repo_collection_id) != 0)
+    g_printerr ("Creating commit with different collection ID (%s) than the repo's one (%s)\n", collection_id, repo_collection_id);
+
+  if (collection_id == NULL || opt_key_ids == NULL)
+    {
+      *out_metadata = (metadata != NULL) ? g_variant_ref (metadata) : NULL;
+      return TRUE;
+    }
+
+  g_autoptr(GVariantBuilder) builder = NULL;
+  builder = ot_util_variant_builder_from_variant (metadata,
+                                                  G_VARIANT_TYPE_VARDICT);
+  g_variant_builder_add (builder, "{s@v}", "ostree.commit.collection-ref",
+                         g_variant_new_variant (g_variant_new ("(ss)",
+                                                               collection_id,
+                                                               opt_branch)));
+
+  *out_metadata = g_variant_builder_end (builder);
+  g_variant_ref_sink (*out_metadata);
```

Comment:
I tend to do `*out_metadata = g_variant_ref_sink (g_variant_builder_end (builder)));`